File: dual.py
import attr
from typing import Union

import numpy as np


Number = Union[float, int, np.ndarray]


@attr.s
class Dual:
    first: Number = attr.ib()
    second: Number = attr.ib()

    @staticmethod
    def _lift(primitive: Number) -> 'Dual':
        return Dual(primitive, 0)

    def __add__(self, other: Union['Dual', Number]) -> 'Dual':
        if isinstance(other, Dual):
            return Dual(self.first + other.first, self.second + other.second)
        else:
            return self + Dual._lift(other)

    def __radd__(self, other: Union['Dual', Number]) -> 'Dual':
        return self + other

    def __sub__(self, other: Union['Dual', Number]) -> 'Dual':
        if isinstance(other, Dual):
            return Dual(self.first - other.first, self.second - other.second)
        else:
            return self - Dual._lift(other)

    def __rsub__(self, other: Union['Dual', Number]) -> 'Dual':
        return self - other

    def __mul__(self, other: Union['Dual', Number]) -> 'Dual':
        if isinstance(other, Dual):
            return Dual(self.first * other.first, self.second * other.second)
        else:
            return self * Dual._lift(other)

    def __rmul__(self, other: Union['Dual', Number]) -> 'Dual':
        return self * other

    def __truediv__(self, other: Union['Dual', Number]) -> 'Dual':
        if isinstance(other, Dual):
            if other.first == 0:
                raise Exception
            first = self.first / other.first
            second = ((self.second * other.first) - (self.first * other.second)) / (other.first ** 2)
            return Dual(first, second)
        else:
            return self / Dual._lift(other)

    def sin(self) -> 'Dual':
        return Dual(np.sin(self.first), self.second * np.cos(self.first))

    def cos(self) -> 'Dual':
        return Dual(np.cos(self.first), -self.second * np.sin(self.first))

    def exp(self) -> 'Dual':
        return Dual(np.exp(self.first), self.second * np.exp(self.first))

    def log(self) -> 'Dual':
        if self.first <= 0:
            raise Exception
        return Dual(np.log(self.first), self.second / self.first)

    def __pow__(self, k: Number) -> 'Dual':
        return Dual(self.first ** k, k * (self.first ** (k - 1)) * self.second)

    def __abs__(self) -> 'Dual':
        return Dual(abs(self.first), self.second * np.sign(self.first))


UNumber = Union[Number, Dual]


def sin(n: UNumber) -> UNumber:
    if isinstance(n, Dual):
        return n.sin()
    else:
        return np.sin(n)


def cos(n: UNumber) -> UNumber:
    if isinstance(n, Dual):
        return n.cos()
    else:
        return np.cos(n)


def exp(n: UNumber) -> UNumber:
    if isinstance(n, Dual):
        return n.exp()
    else:
        return np.exp(n)


def log(n: UNumber) -> UNumber:
    if isinstance(n, Dual):
        return n.log()
    else:
        return np.log(n)

File: nim/dual.nim
import math

type
  Dual[T: SomeNumber] = object
    first: T
    second: T

### Fundamental operations

func add(left, right: Dual): Dual =
  Dual(first: left.first + right.first,
       second: left.second + right.second)

func sub(left, right: Dual): Dual =
  Dual(first: left.first - right.first,
       second: left.second - right.second)

func mul(left, right: Dual): Dual =
  Dual(first: left.first * right.first,
       second: left.second * right.first + left.first * right.second)

func `div`[T: SomeInteger](left, right: Dual[T]): Dual[T] =
  Dual[T](first: left.first div right.first,
          second: ((left.second * right.first) - (left.first * right.second)) div (right.first ^ 2))

func `div`[T: SomeFloat](left, right: Dual[T]): Dual[T] =
  Dual[T](first: left.first / right.first,
          second: ((left.second * right.first) - (left.first * right.second)) / (right.first ^ 2))

func sin[T: SomeNumber](d: Dual[T]): Dual[float] =
  Dual[float](first: float(d.first).sin(),
              second: float(d.second) * float(d.first).cos())

func cos[T: SomeNumber](d: Dual[T]): Dual[float] =
  Dual[float](first: float(d.first).cos(),
              second: float(-d.second) * float(d.first).sin())

func exp[T](d: Dual[T]): Dual[float] =
  let v = float(d.first).exp()
  Dual[float](first: v,
              second: float(d.second) * v)

func ln[T: SomeNumber](d: Dual[T]): Dual[float] =
  Dual[float](first: float(d.first).ln(),
              second: d.second / d.first)

func pow[T: SomeFloat](d: Dual[T], k: T): Dual[T] =
  Dual[T](first: d.first.pow(k),
          second: k * (d.first.pow(k - 1.0)) * d.second)

func pow[T: SomeInteger](d: Dual[T], k: T): Dual[T] =
  Dual[T](first: d.first ^ k,
          second: k * (d.first^ (k - 1)) * d.second)

# TODO pow:
#  - first float, second int
#  - first int, second float

func abs[T](d: Dual[T]): Dual[T] =
  Dual[T](first: d.first.abs(),
          second: T(float(d.second) * float(d.first.sgn())))

### aliases

func `+`(left, right: Dual): Dual =
  left.add(right)

func `-`(left, right: Dual): Dual =
  left.sub(right)

func `*`(left, right: Dual): Dual =
  left.mul(right)

func `/`(left, right: Dual): Dual =
  left.div(right)

func `^`(d: Dual, k: SomeNumber): Dual =
  d.pow(k)


when isMainModule:
  let
    a = Dual[int](first: 2, second: 3)
    b = Dual[int](first: 3, second: 4)
  echo "add: ", a.add(b)
  echo "sub: ", a.sub(b)
  echo "mul: ", a.mul(b)
  echo "div: ", a.div(b)
  echo "sin: ", a.sin()
  echo "cos: ", a.cos()
  echo "exp: ", a.exp()
  echo "ln : ", a.ln()
  echo "pow: ", a.pow(2)
  echo "abs: ", a.abs()

  let
    c = Dual[float](first: 2.0, second: 3.0)
    d = Dual[float](first: 3.0, second: 4.0)
  echo "add: ", c.add(d)
  echo "sub: ", c.sub(d)
  echo "mul: ", c.mul(d)
  echo "div: ", c.div(d)
  echo "sin: ", c.sin()
  echo "cos: ", c.cos()
  echo "exp: ", c.exp()
  echo "ln : ", c.ln()
  echo "pow: ", c.pow(2.0)
  echo "abs: ", c.abs()
